apps/dashboard/app/models/batch_connect/session.rb

```diff
@@ -556,10 +556,14 @@ def connection_in_info?
     # Whether to allow SSH to node running the job
     # @return [Boolean] whether to allow SSH to node running the job
     def ssh_to_compute_node?
-      if !cluster_ssh_to_compute_node?.nil?
+      if cluster_ssh_to_compute_node?.nil? && app_ssh_to_compute_node?.nil?
```

Comment:
Not sure why you're adding the `&& app_ssh_to_compute_node?.nil?` expression here.

This `if/else` gets less specific as it goes down, so it seems like `app_ssh_to_compute_node` is the first thing to check, then you get less specific in each subsequent `elsif` statement.

Reply:
If both are nil - default to global setting
If one is the nil and the other isn't - check the one that isn't nil
If neither is nil, determine with &&

That's my reasoning

Reply:
OK - I kinda get why we're checking for nil here, `nil` being the value of unset so you can check other configs.

I'll have to look this over a few more times, but there's got to be something cleaner here. Maybe temporary variables?

I see that it's a bit complex to keep the precedence + combinations of other configs.

Reply:
Yeah, I see your point for sure. I'll try to clean it up a bit.

Reply:
Yea lol sorry, I'm asking you in one comment and answering myself in another (after actually considering the question). 

Yea I would ask if you can take another swing at this to see if there's something cleaner we can do here.

Reply:
OK - I see that `nil` values here are actually making this harder than it should be.

If everything returned booleans, would this diff have the same logic, precedence and default behavior?

[ssh_form.txt](https://github.com/OSC/ondemand/files/13480431/ssh_form.txt)

Right now, the global config won't ever be nil.  `ood_core` can return nil, but we're adding the third case now.

What do you think about everything returning straight up booleans (we'd have to change ood_core, but I'm fine with that)?

Reply:
I'm afraid I'm not following. If we treat nil for the two configurable options as false we'll never fall back to the global config.

Reply:
I mean if everything returned booleans (and defaulted to `true`), couldn't it just be this?

```ruby
     def ssh_to_compute_node?
       return false unless global_ssh_to_compute_node?
       return false unless cluster_ssh_to_compute_node?
       
       app_ssh_to_compute_node?
     end
```

Reply:
This fails if global is false but app & cluster are true.

Reply:
> This fails if global is false but app & cluster are true.

I guess I'd have to check the logic under test. If it's disabled globally, you shouldn't see it regardless of the other settings. Can it be the test is wrong? I'm not discounting that I'm wrong, but re-reading that, I think that's the right logic.

Reply:
OK, alright, I see this logic isn't correct as we've currently defined _correct_.

Let us consider this use case
* i've disabled it globally
* but i've enabled it on a cluster

What's my expectation as a user here? I think we got it wrong initially. I think as a user, I would expect the global setting to override any specific setting.

I think we need a minute to mull this over, because at this point it's a behavior change, so we need a second to consider what the user expectation is here.

Reply:
OK - 1,000 apologies for all the back and forth while we (I) figure this out.

> Let us consider this use case
>
>   * i've disabled it globally
>   * but i've enabled it on a cluster

I think it's rather simple - If _I_ did that on this system I _would_ expect that to work. And perhaps more importantly, I would be frustrated to find out that I would have to enable it globally and disable on N other clusters to achieve the same behavior.  The situation would be worse if I were trying to enable just 1 app.

So we need that third `nil` values to represent nothing has been configured (I'm fairly sure, I've played it out a few ways with just booleans and I don't think that's possible).

So we're kinda back to where we started - but I think I do like that style, so we should use that style and just capture each use case like when *the app is overriding configs above it* and so on.

```ruby
def ssh_to_compute_node?
  # note these method names are likely too long
  return true if app_override_ssh_to_compute_node? 
  return true if cluster_override_ssh_to_compute_node?

  global_ssh_to_compute_node?
end
```